File: clojure/day07.clj
(ns day07
  (:require aoc
            [clojure.string :as str]
            [clojure.core.match :refer [match]]))


(defn add! [path size sizes]
  (let [val (parse-long size)]
    (loop [sizes sizes
           path path]
      (if (empty? path) sizes
          (recur (assoc! sizes path (+ val (sizes path 0)))
                 (pop path))))))

(defn calc-sizes [commands]
  (->> commands
       (reduce (fn [[sizes path] cmd]
                 (match (str/split cmd #" ")
                   ["$" "cd" ".."] [sizes (pop path)]
                   ["$" "cd" f]    [sizes (conj path f)]
                   (:or ["$" "ls"]
                        ["dir" _]) [sizes path]
                   [size _]        [(add! path size sizes) path]))
               [(transient {}) []])
       first
       persistent!))

(defn part-1 [sizes]
  (->> sizes
       (filter #(<= % 100000))
       (reduce +)))

(defn part-2 [sizes total-size]
  (let [goal 40000000
        to-remove (- total-size goal)]
    (->> sizes
         (filter #(>= % to-remove))
         (reduce min))))

(defn solve
  ([] (solve 7))
  ([input]
   (let [folders    (calc-sizes (aoc/read-input input))
         total-size (folders ["/"])
         sizes      (vals folders)]
     [(part-1 sizes)
      (part-2 sizes total-size)])))


(solve)

File: python/day07.py
from aoc import *
from collections import defaultdict
from itertools import accumulate


def parse_input(data):
    sizes = defaultdict(int)
    for line in data:
        match line.split():
            case '$', 'cd', "/":  path = ['/']
            case '$', 'cd', "..": path.pop()
            case '$', 'cd', f:    path.append(f+'/')
            case '$', 'ls':       pass
            case 'dir', _:        pass
            case val, _:
                for p in accumulate(path):
                    sizes[p] += int(val)
    return sizes


def part_1(sizes):
    return sum(v for v in sizes.values()
               if v <= 100_000)


def part_2(sizes):
    total_size = sizes['/']
    goal = 40_000_000
    to_remove = total_size - goal
    return min(v for v in sizes.values()
               if v >= to_remove)


def solve(file=7):
    sizes = parse_input(read_input(file))
    return part_1(sizes), part_2(sizes)


print(solve())










def test():
    assert solve("07_test") == (95437, 24933642)
    assert solve() == (1444896, 404395)
